File: lib/python/src/cabincrew_protocol/gateway_llm.py
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field

class LLMGatewayRequest(BaseModel):
    """
    A request to an LLM intercepted by the gateway.
    Defined in schemas/draft/llm-gateway.json
    """
    request_id: str
    timestamp: str
    source: Optional[str] = None
    model: str
    provider: Optional[str] = None
    input: Dict[str, Any]
    context: Optional[Dict[str, Any]] = None

class GatewayApproval(BaseModel):
    approval_id: Optional[str] = None
    required_role: Optional[str] = None
    reason: Optional[str] = None

class LLMGatewayResponse(BaseModel):
    """
    Decision produced by the LLM gateway.
    Defined in schemas/draft/llm-gateway.json
    """
    request_id: str
    timestamp: str
    decision: str = Field(..., description="allow, warn, deny, require_approval")
    warnings: Optional[List[str]] = None
    violations: Optional[List[str]] = None
    approval: Optional[GatewayApproval] = None
    routed_model: Optional[str] = None
    rewritten_input: Optional[Dict[str, Any]] = None
    gateway_payload: Optional[Dict[str, Any]] = None

class LLMGatewayRule(BaseModel):
    match: Dict[str, Any]
    action: str
    metadata: Optional[Dict[str, Any]] = None

class LLMGatewayPolicyConfig(BaseModel):
    """
    OPA + ONNX policy configuration for the LLM Gateway.
    Defined in schemas/draft/llm-gateway.json
    """
    opa_policies: Optional[List[str]] = None
    onnx_models: Optional[List[str]] = None
    model_routing: Optional[Dict[str, Any]] = None
    rules: Optional[List[LLMGatewayRule]] = None

File: lib/go/cabincrew/gateway_llm.go
package cabincrew

// LLMGatewayRequest is a request to an LLM intercepted by the gateway.
// Defined in schemas/draft/llm-gateway.json
type LLMGatewayRequest struct {
	RequestID string                 `json:"request_id"`
	Timestamp string                 `json:"timestamp"`
	Source    string                 `json:"source,omitempty"`
	Model     string                 `json:"model"`
	Provider  string                 `json:"provider,omitempty"`
	Input     map[string]interface{} `json:"input"`
	Context   map[string]interface{} `json:"context,omitempty"`
}

// LLMGatewayResponse is a decision produced by the LLM gateway.
// Defined in schemas/draft/llm-gateway.json
type LLMGatewayResponse struct {
	RequestID      string                 `json:"request_id"`
	Timestamp      string                 `json:"timestamp"`
	Decision       string                 `json:"decision"` // allow, warn, deny, require_approval
	Warnings       []string               `json:"warnings,omitempty"`
	Violations     []string               `json:"violations,omitempty"`
	Approval       *GatewayApproval       `json:"approval,omitempty"`
	RoutedModel    string                 `json:"routed_model,omitempty"`
	RewrittenInput map[string]interface{} `json:"rewritten_input,omitempty"`
	GatewayPayload map[string]interface{} `json:"gateway_payload,omitempty"`
}

type GatewayApproval struct {
	ApprovalID   string `json:"approval_id,omitempty"`
	RequiredRole string `json:"required_role,omitempty"`
	Reason       string `json:"reason,omitempty"`
}

// LLMGatewayPolicyConfig is the OPA + ONNX policy configuration for the LLM Gateway.
// Defined in schemas/draft/llm-gateway.json
type LLMGatewayPolicyConfig struct {
	OpaPolicies  []string               `json:"opa_policies,omitempty"`
	OnnxModels   []string               `json:"onnx_models,omitempty"`
	ModelRouting map[string]interface{} `json:"model_routing,omitempty"`
	Rules        []LLMGatewayRule       `json:"rules,omitempty"`
}

type LLMGatewayRule struct {
	Match    map[string]interface{} `json:"match"`
	Action   string                 `json:"action"`
	Metadata map[string]interface{} `json:"metadata,omitempty"`
}
